### PR TITLE
Allow polyfill providers to specify custom `@babel/runtime` pkg

### LINKED ISSUE #15572: [Bug]: @babel/plugin-transform-runtime not transform Object.hasOwn correctly
### 💻

- [X] Would you like to work on a fix?

### How are you using Babel?

`@babel/cli`

### Input code

```js
const _ = function () {
    const obj = { a: 1}
    console.log(Object.hasOwn(obj, "b"));
    [].includes(1)
      
    const a = new Promise()
    return a
}
export default  _

console.log(_())
```

### Configuration file name

`babel.config.json`

### Configuration

```json
{
  "plugins": [
    [
      "@babel/plugin-transform-runtime",
      {
        "corejs": 3
      }
    ]
  ]
}
```

### Current and expected behavior

I used @babel/cli directly with the command `./node_modules/.bin/babel index.js --out-dir lib`. This is the output:
```js
import _includesInstanceProperty from "@babel/runtime-corejs3/core-js-stable/instance/includes";
import _Promise from "@babel/runtime-corejs3/core-js-stable/promise";
const _ = function () {
  var _context;
  const obj = {
    a: 1
  };
  console.log(Object.hasOwn(obj, "b"));
  _includesInstanceProperty(_context = []).call(_context, 1);
  const a = new _Promise();
  return a;
};
export default _;
console.log(_());
```
The `Promise` and `Array.includes` are replace with `@babel/runtime-corejs3` correctly. But `Object.hasOwn` not. And I find the `@babel/runtime-corejs3/core-js-stable` didn't exist the polyfill of `Object.hasOwn` which is realized in corejs. I have read the doc about `@babel/plugin-transform-runtime`, but I haven't found anything about this. So this behavior is designed so? Or I have missed some informations?

### Environment

```
  System:
    OS: macOS 13.2
  Binaries:
    Node: 16.19.1 - ~/.nvm/versions/node/v16.19.1/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.19.1/bin/yarn
    npm: 8.19.3 - ~/.nvm/versions/node/v16.19.1/bin/npm
  npmPackages:
    @babel/cli: ^7.21.0 => 7.21.0
    @babel/core: ^7.21.4 => 7.21.4
    @babel/plugin-transform-runtime: ^7.21.4 => 7.21.4
    @babel/preset-env: ^7.21.4 => 7.21.4
    @babel/runtime: ^7.21.0 => 7.21.0
    @babel/runtime-corejs3: ^7.21.0 => 7.21.0
```

### Possible solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Closes #15426, fixes https://github.com/babel/babel-polyfills/issues/128#issuecomment-1118917890 (point 1), fixes #15572
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This PR allows polyfill providers to specify their own alternative to `@babel/runtime`, instead of hard-coding `@babel/runtime-corejs2` and `@babel/runtime-corejs3` in `@babel/plugin-transform-runtime`. It is a much better alternative to #15426, since it doesn't require extra configuration from users.

As copilot notices below, most of the changes in this PR are actually just extracting code to a new `polyfills.ts` file so that all the corejs/regenerator logic is self-contained, and can be more easily removed in the future.

The parallel PR in the `babel-polyfills` repository is https://github.com/babel/babel-polyfills/pull/160, and this PR's tests will fail until that one is released.

I'm marking this as internal, since _technically_ is a new feature for `@babel/plugin-transform-runtime` but in practice it only changes how we internally communicate options between the various plugins.

---

_EDIT by me: this was fun, but it's also not what I needed_ 😂

### Summary
🚚🔌🆕

Refactor the polyfill logic of `babel-plugin-transform-runtime` into a separate module. This makes the main plugin code cleaner and more maintainable, and fixes the regenerator import path.

> _To make the main plugin more neat_
> _They moved the polyfill logic to `polyfills.ts`_
> _With core-js and regenerator in tow_
> _They used a secret key for some options, you know_
> _And updated the helper import to the latest_

### Walkthrough
*  Extract core-js and regenerator polyfill plugins to a separate module `polyfills.ts` ([link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-6b555d8887bc858212b04534b4146442f8880fcf26df080baab49b90b1c10fb4L3-R8), [link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-6b555d8887bc858212b04534b4146442f8880fcf26df080baab49b90b1c10fb4L35-R27), [link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-6b555d8887bc858212b04534b4146442f8880fcf26df080baab49b90b1c10fb4L76-L107), [link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-6b555d8887bc858212b04534b4146442f8880fcf26df080baab49b90b1c10fb4L186-R133), [link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-4efcea8b594912e957802308d04ec85dcdd5260a06f809899a54bc939b0c16a8R1-R138))
*  Simplify `helperGenerator` function by using lazy evaluation of `modulePath` instead of `moduleName` ([link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-6b555d8887bc858212b04534b4146442f8880fcf26df080baab49b90b1c10fb4L186-R133))
*  Update import path of regenerator helper to `@babel/helpers/regenerator` ([link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-6b555d8887bc858212b04534b4146442f8880fcf26df080baab49b90b1c10fb4L290-R143))
*  Use a secret key to pass compatibility options to polyfill plugins ([link](https://github.com/babel/babel/pull/15531/files?diff=unified&w=0#diff-4efcea8b594912e957802308d04ec85dcdd5260a06f809899a54bc939b0c16a8R1-R138))

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15531"></a>

### CODE REVIEW

**Comment by nicolo-ribaudo** [comment]
@JLHwung @liuxingbaoyu Any opinion regarding this?

**Review by JLHwung — APPROVED** [review]

File: packages/babel-plugin-transform-runtime/test/fixtures/regression/with-env-useBuiltIns/output.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+var _createClass = require("@babel/runtime/helpers/createClass");
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-runtime/test/fixtures/regression/with-env-useBuiltIns/output.js:1]: This test is to make sure that when using `preset-env` to load polyfills, and when using `transform-runtime` just for helpers, the helpers are loaded from the package without polyfills (for backwards compatibility).

**Comment by nicolo-ribaudo** [comment]
The failure on older Node.js versions is because they do not resolve symlinks:
```js
   Expected /home/runner/work/babel/babel/packages/babel-plugin-transform-runtime/test/fixtures/absoluteRuntime/true-corejs3-proposals/output.js to match transform output.
    To autogenerate a passing version of this file, delete the file and re-run the tests.

    Diff:

    - Expected
    + Received

    @@ -1,6 +1,6 @@
    - var _mapInstanceProperty = require("<CWD>/packages/babel-runtime-corejs3/core-js/instance/map.js");
    + var _mapInstanceProperty = require("<CWD>/node_modules/@babel/runtime-corejs3/core-js/instance/map.js");
      var _marked = /*#__PURE__*/babelHelpers.regeneratorRuntime().mark(makeIterator);
      _mapInstanceProperty(Array);
```

I'm not sure about what in this PR causes that to change, I will investigate (I don't think it's a bug, but it's annoying to have this difference).

**Review by JLHwung — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
I'm merging this PR because I need to rebase https://github.com/babel/babel/pull/15633 on top of this, to ensure that the `SuppressedError` polyfill is injected in our polyfilled helpers.

**Pull request merged**